mllib/src/main/scala/org/apache/spark/ml/feature/StringIndexer.scala

```diff
@@ -217,69 +289,94 @@ class StringIndexerModel (
           labelToIndex.contains(label)
         }
-        (dataset.na.drop(Array($(inputCol))).where(filterer(dataset($(inputCol)))), false)
-      case _ => (dataset, getHandleInvalid == StringIndexer.KEEP_INVALID)
+        filteredDataset = filteredDataset.where(filterer(dataset(inputColName)))
+      }
     }
 
-    val indexer = udf { label: String =>
-      if (label == null) {
-        if (keepInvalid) {
-          labels.length
-        } else {
-          throw new SparkException("StringIndexer encountered NULL value. To handle or skip " +
-            "NULLS, try setting StringIndexer.handleInvalid.")
-        }
+    for (i <- 0 until outputColNames.length) {
+      val inputColName = inputColNames(i)
+      val outputColName = outputColNames(i)
+      val labelToIndex = labelToIndexArray(i)
+      val labels = labelsArray(i)
+
+      if (!dataset.schema.fieldNames.contains(inputColName)) {
+        logInfo(s"Input column ${inputColName} does not exist during transformation. " +
+          "Skip this column StringIndexerModel transform.")
+        outputColNames(i) = null
       } else {
-        if (labelToIndex.contains(label)) {
-          labelToIndex(label)
-        } else if (keepInvalid) {
-          labels.length
-        } else {
-          throw new SparkException(s"Unseen label: $label.  To handle unseen labels, " +
-            s"set Param handleInvalid to ${StringIndexer.KEEP_INVALID}.")
+        val filteredLabels = getHandleInvalid match {
+          case StringIndexer.KEEP_INVALID => labelsArray(i) :+ "__unknown"
+          case _ => labelsArray(i)
         }
-      }
-    }.asNondeterministic()
 
-    filteredDataset.select(col("*"),
-      indexer(dataset($(inputCol)).cast(StringType)).as($(outputCol), metadata))
+        val metadata = NominalAttribute.defaultAttr
+          .withName(outputColName).withValues(filteredLabels).toMetadata()
+
+        val keepInvalid = (getHandleInvalid == StringIndexer.KEEP_INVALID)
+
+        val indexer = udf { label: String =>
+          if (label == null) {
+            if (keepInvalid) {
+              labels.length
+            } else {
+              throw new SparkException("StringIndexer encountered NULL value. To handle or skip " +
+                "NULLS, try setting StringIndexer.handleInvalid.")
+            }
+          } else {
+            if (labelToIndex.contains(label)) {
+              labelToIndex(label)
+            } else if (keepInvalid) {
+              labels.length
+            } else {
+              throw new SparkException(s"Unseen label: $label.  To handle unseen labels, " +
+                s"set Param handleInvalid to ${StringIndexer.KEEP_INVALID}.")
+            }
+          }
+        }.asNondeterministic()
+
+        outputColumns(i) = indexer(dataset(inputColName).cast(StringType))
+          .as(outputColName, metadata)
+      }
+    }
+    filteredDataset.withColumns(outputColNames.filter(_ != null),
+      outputColumns.filter(_ != null))
```

Comment:
In case `outputColNames` and `outputColNames` are empty, `withColumns` might return an empty dataset, not original dataset.